File: 054. Spiral Matrix/spiralOrder.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
    if (matrix == []) {
        return []
    }
    else {
        const [row,column] = [matrix.length, matrix[0].length]
        var [up,down,left,right] = [0,row-1,0,column-1]
        var res = []
        while (true) {
            for (let j=left;j<=right;j++) {
                res.push(matrix[up][j])
            }                
            up += 1
            if (up > down) {break}
            
            for (let i=up;i<=down;i++) {
                res.push(matrix[i][right])
            }
            right -= 1
            if (right < left) {break}
                
            for (let j=right;j>=left;j--) {
                res.push(matrix[down][j])
            }
            down -= 1
            if (up > down) {break}

            for (let i=down;i>=up;i--) {
                res.push(matrix[i][left])
            }
            left += 1
            if (right < left) {break}
        }
        return res
    }
};

// Runtime 58 ms / Memory 41 MB
// debug: https://jsfiddle.net/

File: 054. Spiral Matrix/spiralOrder.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int row = matrix.size();
        int col = matrix[0].size();
        vector<int> output;
        int U=0, D=row-1, L=0, R=col-1;
        while (true) {
            // go right
            for (int x=L; x<=R; x++) { output.push_back(matrix[U][x]); }
            U++; //change boundry cause visited
            if (U>D) { break; } //boundry condition
            // go down
            for (int y=U; y<=D; y++) { output.push_back(matrix[y][R]); }
            R--;
            if (R<L) { break; }
            // go left
            for (int x=R; x>=L; x--) { output.push_back(matrix[D][x]); }
            D--;
            if (D<U) { break; }
            // go up
            for (int y=D; y>=U; y--) { output.push_back(matrix[y][L]); }
            L++;
            if (L>R) { break; }
        }

        return output;
    }
};